<a href="https://colab.research.google.com/github/alessandranerylima/data_science/blob/main/Data_Cleaning_of_Glassdor_Job_Posting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **1. Project Background:**

The Data Science field has experienced tremendous growth over the past decade. Companies across various industries, including technology, finance, healthcare, and retail, are increasingly leveraging data to make informed decisions. As a result, Data Professionals, who can turn raw data into actionable insights, have become one of the most in-demand professionals globally.

Within this expansive field, a myriad of titles and positions exist, reflecting a spectrum of responsibilities. Examples include Data Analysts, Data Scientists, Data Engineers, and Machine Learning Scientists. While this project does not intend to clarify the precise definitions of these roles, it aims to provide a comprehensive snapshot of the current Data Science job market, as seen through the job postings on Glassdoor.

Additionally, I hope that the findings of this project will not only assist me but also aid other job seekers in uncovering valuable insights for navigating the data science job market


# **2. Primary Purpose:**

This project works with cleaning and trasformation data of Data Science job postings from Glassdoor. Before to make exploratory data analysis and uncover key insights into the job market, it´s crucial to have raw data cleaned and organized.
Below is a list of what will be discussed to organize the data:
  
1.   Transforming Salary Estimate into integers
2.   Extracting out informations of Job Description column
3.   Removing out numbers of Company name column
4.   Rating Column - removing negative rating
5.   Location column - extracting abbreviation of the state
6.   Drop unwanted column


# **3. Data set:**

This is a dataset of data science job posts in glassdoor - one of the world's largest job and recruiting sites.

The dataset contains 672 observations and 14 variables, each observation represents a unique job posting for a Data Science-related position on Glassdoor.

The content of  data was scrapped from glassdoor's website and cosists in a  uncleaned version.

The variables provide detailed information about each job posting. The columns explanation are as follows:

1. Job Title: Title of the job posting
2. Salary Estimation: Salary range for that particular job
3. Job Description: This contains the full description of that job
4. Rating: Rating of that post
5. Company: Name of company
6. Location: Location of the company
7. Headquarter: Location of the headquater
8. Size: Total employee in that company
9. Founded
10. Type of ownership: Describes the company type i.e non-profit/public/private farm etc
11. Industry
12. Sector: Field applicant will work in
13. Revenue: Total revenue of the company
14. Competitors


Link: https://www.kaggle.com/datasets/rashikrahmanpritom/data-science-job-posting-on-glassdoor?select=Uncleaned_DS_jobs.csv

# **4. Loading the datas**

In [1]:
# import packages
import pandas as pd
import numpy as np




In [2]:
df = pd.read_csv(r'/content/Uncleaned_DS_jobs.csv')

# **5. Understanding the datas**

In [3]:
# Columns of the data set
df.columns

Index(['index', 'Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors'],
      dtype='object')

In [4]:
#Checking data type of each column
df.dtypes

index                  int64
Job Title             object
Salary Estimate       object
Job Description       object
Rating               float64
Company Name          object
Location              object
Headquarters          object
Size                  object
Founded                int64
Type of ownership     object
Industry              object
Sector                object
Revenue               object
Competitors           object
dtype: object

In [5]:
#Checking data shape
df.shape

(672, 15)

In [6]:
# Statistical summary of the column Rating
df['Rating'].describe()

count    672.000000
mean       3.518601
std        1.410329
min       -1.000000
25%        3.300000
50%        3.800000
75%        4.300000
max        5.000000
Name: Rating, dtype: float64

In [7]:
#Checking for duplicate records
df[df.duplicated()]

Empty DataFrame
Columns: [index, Job Title, Salary Estimate, Job Description, Rating, Company Name, Location, Headquarters, Size, Founded, Type of ownership, Industry, Sector, Revenue, Competitors]
Index: []

In [8]:
# Checking for missing values
df.isnull().sum()

index                0
Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         0
Location             0
Headquarters         0
Size                 0
Founded              0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
Competitors          0
dtype: int64

In [9]:
df.head()

index          Job Title               Salary Estimate  \
0      0  Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1      1     Data Scientist  $137K-$171K (Glassdoor est.)   
2      2     Data Scientist  $137K-$171K (Glassdoor est.)   
3      3     Data Scientist  $137K-$171K (Glassdoor est.)   
4      4     Data Scientist  $137K-$171K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  
0            EmblemHealth, UnitedHealth Group, Aetna  
1                                                 -1  
2                                                 -1  
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4               Commerce Signals, Cardlytics, Yodlee

# **6. Preparing and Cleaning the datas**

1- Transforming Salary Estimate into integers

The "salary estimate" column contains strings with upper and lower limit estimates provided by Glassdoor or employers. For exploratory analysis, it is necessary to extract these lower and upper limits and convert them into integers. Nevertheless, before proceeding with this data transformation, it´s crucial examine the format of 'Salary Estimate' column in order to ensure there are any unexpected values.

In [10]:
# extract the column
salary = df['Salary Estimate']
print(salary.head(10), '\n'*1)

0    $137K-$171K (Glassdoor est.)
1    $137K-$171K (Glassdoor est.)
2    $137K-$171K (Glassdoor est.)
3    $137K-$171K (Glassdoor est.)
4    $137K-$171K (Glassdoor est.)
5    $137K-$171K (Glassdoor est.)
6    $137K-$171K (Glassdoor est.)
7    $137K-$171K (Glassdoor est.)
8    $137K-$171K (Glassdoor est.)
9    $137K-$171K (Glassdoor est.)
Name: Salary Estimate, dtype: object 



In [11]:
# there are observations do not contain 'Glassdoor est',
# instead, they contain 'Employer est.'
print(salary[~salary.str.contains('(Glassdoor est.)', regex = False)])

303    $145K-$225K(Employer est.)
304    $145K-$225K(Employer est.)
305    $145K-$225K(Employer est.)
306    $145K-$225K(Employer est.)
307    $145K-$225K(Employer est.)
308    $145K-$225K(Employer est.)
309    $145K-$225K(Employer est.)
310    $145K-$225K(Employer est.)
311    $145K-$225K(Employer est.)
312    $145K-$225K(Employer est.)
313    $145K-$225K(Employer est.)
314    $145K-$225K(Employer est.)
315    $145K-$225K(Employer est.)
316    $145K-$225K(Employer est.)
317    $145K-$225K(Employer est.)
318    $145K-$225K(Employer est.)
319    $145K-$225K(Employer est.)
320    $145K-$225K(Employer est.)
321    $145K-$225K(Employer est.)
322    $145K-$225K(Employer est.)
Name: Salary Estimate, dtype: object


In [12]:
# check if that all values have the unit of K/thousand
salary[~salary.str.contains('K')].shape[0]

0

In [13]:
# creating three new columns salary estimation's: upper limit, lower limit and midpoint
# and removing non-numerical characters

df['Salary Estimate']=df['Salary Estimate'].str.strip('(Employer est.)K$(Glassdoor est.)')
df[['salary_lower','salary_upper']]=df['Salary Estimate'].str.split('-',1,expand=True)
df['salary_lower']=df['salary_lower'].str.strip('K')
df['salary_upper']=df['salary_upper'].str.strip('$')
df['salary_midpoint'] = (df.salary_lower.astype('int') +
                           df.salary_upper.astype('int')) / 2


<ipython-input-13-3a361e18db3f>:5: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.
  df[['salary_lower','salary_upper']]=df['Salary Estimate'].str.split('-',1,expand=True)


In [14]:
#Display the dataframe with the transformation applied
df

index          Job Title Salary Estimate  \
0        0  Sr Data Scientist       137K-$171   
1        1     Data Scientist       137K-$171   
2        2     Data Scientist       137K-$171   
3        3     Data Scientist       137K-$171   
4        4     Data Scientist       137K-$171   
..     ...                ...             ...   
667    667     Data Scientist       105K-$167   
668    668     Data Scientist       105K-$167   
669    669     Data Scientist       105K-$167   
670    670     Data Scientist       105K-$167   
671    671     Data Scientist       105K-$167   

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
667  Summary\n\nWe’re looking for a data scientist ...     3.6   
668  Job Description\nBecome a thought leader withi...    -1.0   
669  Join a thriving company that is changing the w...    -1.0   
670  100 Remote Opportunity As an AINLP Data Scient...     5.0   
671  Description\n\nThe Data Scientist will be part...     2.7   

                  Company Name           Location            Headquarters  \
0             Healthfirst\n3.1       New York, NY            New York, NY   
1                 ManTech\n4.2      Chantilly, VA             Herndon, VA   
2          Analysis Group\n3.8         Boston, MA              Boston, MA   
3                 INFICON\n3.5         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions\n2.9       New York, NY            New York, NY   
..                         ...                ...                     ...   
667              TRANZACT\n3.6       Fort Lee, NJ            Fort Lee, NJ   
668                       JKGT  San Francisco, CA                      -1   
669                 AccessHope      Irwindale, CA                      -1   
670  ChaTeck Incorporated\n5.0  San Francisco, CA         Santa Clara, CA   
671         1-800-Flowers\n2.7       New York, NY         Carle Place, NY   

                        Size  Founded        Type of ownership  \
0     1001 to 5000 employees     1993   Nonprofit Organization   
1    5001 to 10000 employees     1968         Company - Public   
2     1001 to 5000 employees     1981  Private Practice / Firm   
3      501 to 1000 employees     2000         Company - Public   
4        51 to 200 employees     1998        Company - Private   
..                       ...      ...                      ...   
667   1001 to 5000 employees     1989        Company - Private   
668                       -1       -1                       -1   
669                       -1       -1                       -1   
670        1 to 50 employees       -1        Company - Private   
671   1001 to 5000 employees     1976         Company - Public   

                                  Industry             Sector  \
0                       Insurance Carriers          Insurance   
1                   Research & Development  Business Services   
2                               Consulting  Business Services   
3    Electrical & Electronic Manufacturing      Manufacturing   
4                  Advertising & Marketing  Business Services   
..                                     ...                ...   
667                Advertising & Marketing  Business Services   
668                                     -1                 -1   
669                                     -1                 -1   
670                Advertising & Marketing  Business Services   
671                              Wholesale  Business Services   

                        Revenue  \
0      Unknown / Non-Applicable   
1        $1 to $2 billion (USD)   
2    $100 to $500 m

2- Extracting out informations of Job Description column

In [15]:
# record the original title just in case
df['old title'] = df['Job Title']

# create a column to record whether a title has been integrated
df['integrated title'] = pd.Series(np.zeros(len(df)), dtype = int)

# remove whitespace, change into lower cases
df['Job Title'] = df['Job Title'].str.strip().str.lower()

# check unique values of the titles and their counts
title_counts = df['Job Title'].value_counts()
print(title_counts)



data scientist                                            337
data engineer                                              26
senior data scientist                                      19
machine learning engineer                                  16
data analyst                                               12
                                                         ... 
data science instructor                                     1
business data analyst                                       1
purification scientist                                      1
data engineer, enterprise analytics                         1
ai/ml - machine learning scientist, siri understanding      1
Name: Job Title, Length: 172, dtype: int64


In [16]:
# get the count of different occurance frequencies
title_occurance_freq = title_counts.value_counts().sort_index().to_frame()
title_occurance_freq.index.name = 'occurance frequency'
title_occurance_freq.columns = ['count']
print(title_occurance_freq, '\n'*2)

print('Titles that have an occurance of at least 5 times cover ',
      (title_counts[title_counts >= 5].sum())/df.shape[0]*100,
      '% of the raw data')

                     count
occurance frequency       
1                       96
2                       55
3                       11
4                        3
5                        1
6                        1
12                       1
16                       1
19                       1
26                       1
337                      1 


Titles that have an occurance of at least 5 times cover  62.648809523809526 % of the raw data


The raw dataset contains 172 unique titles; however, 96 of these titles appear only once, 55 appear twice, and titles occurring at least five times comprise approximately 62% of our raw data. To ensure sufficient data for each title we analyze, we will retain only those titles that occur five times or more, and we will consolidate the remaining titles into these more frequently occurring categories.

To accomplish this, we will first assess the similarities between the titles that appear less than five times and those that appear at least five times. Based on this assessment, we will then group similar titles under a common title using the "fuzzywuzzy" library in Python.

The "fuzzywuzzy" library is a useful tool for performing string comparisons based on string matching. It uses the Levenshtein distance (also known as edit distance) to calculate the similarity between two strings. "Fuzzywuzzy" is commonly used in tasks such as data deduplication, record matching, and string sorting based on its similarity.

In [17]:
pip install fuzzywuzzy


In [18]:
pip install --upgrade pip


In [19]:
import fuzzywuzzy
from fuzzywuzzy import process

/usr/local/lib/python3.10/dist-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [20]:
# Now we use the extract function from fuzzywuzzy to assess the title
# similarity, it allows you to input a title, and the algorithm will
# find the top n most similar titles each with a similar score.
titles_show_up_5 = title_counts[title_counts >= 5].index.to_list()

for i in titles_show_up_5:
    print('These are the similar titles to ', i, '\n',
          fuzzywuzzy.process.extract(i, df['Job Title'].unique(),
                                     limit = 20,
                                     scorer = fuzzywuzzy.fuzz.token_sort_ratio),
          '\n'*2)

These are the similar titles to  data scientist 
 [('data scientist', 100), ('sr data scientist', 90), ('(sr.) data scientist -', 90), ('ai data scientist', 90), ('sr. data scientist', 90), ('data scientist - risk', 85), ('lead data scientist', 85), ('data scientist 3 (718)', 82), ('staff data scientist', 82), ('sr. data scientist ii', 82), ('senior data scientist', 80), ('data scientist (ts/sci)', 80), ('ai ops data scientist', 80), ('data scientist - contract', 76), ('associate data scientist', 74), ('principal data scientist', 74), ('ngs scientist', 74), ('geospatial data scientist', 72), ('experienced data scientist', 70), ('human factors scientist', 70)] 


These are the similar titles to  data engineer 
 [('data engineer', 100), ('jr. data engineer', 90), ('big data engineer', 87), ('data engineer - kafka', 81), ('data engineer (remote)', 79), ('senior data engineer', 79), ('data analyst/engineer', 76), ('software data engineer', 74), ('data analytics engineer', 72), ('cloud data

We observe that most similar titles differ primarily in their wording or formatting, such as 'Sr. Data Analyst', 'Senior Data Analyst', and 'Experienced Data Analyst'.

To streamline these variations, we propose integrating titles that appear fewer than 5 times using a series of conditional checks. Furthermore, we've identified several management-level titles that don't align with the most frequent titles. To address this, we plan to create a new category, 'Data Science Manager', to encompass all manager-level titles.

In [21]:
# Start to integrate the titles that appear less than 5 times, which
# means we will only keep the titles below and integrate others into them.
titles_we_keep = titles_show_up_5.copy()
titles_we_keep.append('data science manager')
titles_we_keep

['data scientist',
 'data engineer',
 'senior data scientist',
 'machine learning engineer',
 'data analyst',
 'senior data analyst',
 'senior data engineer',
 'data science manager']

In [22]:
# Now we use a series of if-else check to consolidate the titles. To do that
# we start by naming some key words to identify certain titles.

senior = ['senior', 'sr', 'experienced', 'ii', 'iii', 'staff']
manager = ['manager', 'management', 'lead', 'principal', 'director',
           'president', 'vp']

for i in range(0, len(df)):
    if df.loc[i, 'Job Title'] not in titles_we_keep:
        df.loc[i, 'integrated title'] = 1
        title = df.loc[i, 'Job Title']

        if any(key in title for key in manager):
            df.loc[i, 'Job Title'] = 'data science manager'
        elif 'machine learning' in title:
            df.loc[i, 'Job Title'] = 'machine learning engineer'
        elif any(key in title for key in senior):
            if 'engineer' in title:
                df.loc[i, 'Job Title'] = 'senior data engineer'
            elif 'analyst' in title:
                df.loc[i, 'Job Title'] = 'senior data analyst'
            elif 'scientist' in title:
                df.loc[i, 'Job Title'] = 'senior data scientist'
        elif 'engineer' in title:
            df.loc[i, 'Job Title'] = 'data engineer'
        elif 'analyst' in title:
            df.loc[i, 'Job Title'] = 'data analyst'
        elif 'scientist' in title:
            df.loc[i, 'Job Title'] = 'data scientist'

In [23]:
# There are still some titles with unique wordings that are not integrated,
# we can integrate using the following conditional checks.
titles_need_change_manually = df['Job Title'].value_counts().index.to_list()[8:]

for i in range(0, len(df)):
    if df.loc[i, 'Job Title'] in titles_need_change_manually:
        df.loc[i, 'integrated title'] = 1
        if ('environmental data science' in title or
           'it partner digital health technology and data science' in title):
            df.loc[i, 'Job Title']  = 'data analyst'
        else:
            df.loc[i, 'Job Title']  = 'data scientist'

# see results
df['Job Title'].value_counts()

data scientist               421
data engineer                 63
senior data scientist         51
machine learning engineer     45
data analyst                  35
data science manager          30
senior data analyst           20
senior data engineer           7
Name: Job Title, dtype: int64

We observe that the conditional checks have effectively managed most of the titles that appear only once or twice, and we further refined the integration manually for the unusual cases. Ultimately, we consolidated all the titles into the 8 common categories we have chosen. For those who wish to validate the process, the following code provides a comparison between the original titles and the integrated ones.

In [24]:
#pd.set_option('display.max_rows', None)
#data.loc[df['integrated title'] == 1,  ['Job Title', 'old title']]

3- Removing out numbers of Company name column:

For the company name column, the biggest problem we can notice is that each name comes with its rating, which is already recorded in another column. So, we need to get rid of the unnessary postfix

In [25]:
df['Company Name'] = df['Company Name'].str.split('\n').str[0]

In [26]:
#Display the dataframe with the transformation applied
df.head(10)

index                  Job Title Salary Estimate  \
0      0      senior data scientist       137K-$171   
1      1             data scientist       137K-$171   
2      2             data scientist       137K-$171   
3      3             data scientist       137K-$171   
4      4             data scientist       137K-$171   
5      5             data scientist       137K-$171   
6      6  machine learning engineer       137K-$171   
7      7             data scientist       137K-$171   
8      8      senior data scientist       137K-$171   
9      9             data scientist       137K-$171   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   
5  About Us:\n\nHeadquartered in beautiful Santa ...     4.2   
6  Posting Title\nData Scientist / Machine Learni...     3.9   
7  Introduction\n\nHave you always wanted to run ...     3.5   
8  Intuit is seeking a Staff Data Scientist to co...     4.4   
9  Ready to write the best chapter of your career...     3.6   

         Company Name           Location            Headquarters  \
0         Healthfirst       New York, NY            New York, NY   
1             ManTech      Chantilly, VA             Herndon, VA   
2      Analysis Group         Boston, MA              Boston, MA   
3             INFICON         Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions       New York, NY            New York, NY   
5         HG Insights  Santa Barbara, CA       Santa Barbara, CA   
6            Novartis      Cambridge, MA      Basel, Switzerland   
7              iRobot        Bedford, MA             Bedford, MA   
8       Intuit - Data      San Diego, CA       Mountain View, CA   
9  XSELL Technologies        Chicago, IL             Chicago, IL   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   
5      51 to 200 employees     2010        Company - Private   
6         10000+ employees     1996         Company - Public   
7   1001 to 5000 employees     1990         Company - Public   
8  5001 to 10000 employees     1983         Company - Public   
9      51 to 200 employees     2014        Company - Private   

                                   Industry                     Sector  \
0                        Insurance Carriers                  Insurance   
1                    Research & Development          Business Services   
2                                Consulting          Business Services   
3     Electrical & Electronic Manufacturing              Manufacturing   
4                   Advertising & Marketing          Business Services   
5              Computer Hardware & Software     Information Technology   
6                 Biotech & Pharmaceuticals  Biotech & Pharmaceuticals   
7  Consumer Electronics & Appliances Stores                     Retail   
8              Computer Hardware & Software     Information Technology   
9   Enterprise Software & Network Solutions     Information Technology   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   
5    Unknown / Non-Applicable   
6          $10+ billion (USD)   
7      $1 to $2 billion (USD)   
8      $2 to $5 billion (USD)   
9    Unknown / Non-Applicable   

                                         Competitors salary

4- Rating Column - removing negative rating

In [27]:
df.sort_values(by='Rating', ascending=True)

index                  Job Title Salary Estimate  \
457    457             data scientist        69K-$116   
425    425             data scientist       124K-$198   
351    351             data scientist       122K-$146   
504    504             data scientist        95K-$119   
500    500             data scientist        95K-$119   
..     ...                        ...             ...   
209    209             data scientist        79K-$106   
405    405  machine learning engineer       110K-$163   
113    113              data engineer        99K-$132   
355    355             data scientist       122K-$146   
118    118             data scientist        99K-$132   

                                       Job Description  Rating  \
457  Work in a fast growing startup with unlimited ...    -1.0   
425  Job Description\nSelecting features, building ...    -1.0   
351  About Our AI/ML Team\n\nOur mission is to buil...    -1.0   
504  Job Description\nWorking at Sophinea\n\nSophin...    -1.0   
500  Job Overview: The Data Scientist is a key memb...    -1.0   
..                                                 ...     ...   
209  US Citizenship Required and (TS/SCI with FSP o...     5.0   
405  We are looking for an experienced engineer wit...     5.0   
113  At Phantom AI, experience the fast paced envir...     5.0   
355  US Citizenship Required and (TS/SCI with FSP o...     5.0   
118  ArsenalBio’s mission is to develop efficacious...     5.0   

                 Company Name                 Location  \
457             Stride Search        San Francisco, CA   
425              Microagility             New York, NY   
351          Point72 Ventures            Palo Alto, CA   
504                  Sophinea            Chantilly, VA   
500            Hatch Data Inc        San Francisco, CA   
..                        ...                      ...   
209  Phoenix Operations Group   Annapolis Junction, MD   
405                  LifeOmic              Raleigh, NC   
113                Phantom AI           Burlingame, CA   
355  Phoenix Operations Group   Annapolis Junction, MD   
118       Arsenal Biosciences  South San Francisco, CA   

                Headquarters                 Size  Founded  Type of ownership  \
457     Westlake Village, CA    1 to 50 employees       -1  Company - Private   
425            Princeton, NJ    1 to 50 employees       -1  Company - Private   
351                       -1                   -1       -1                 -1   
504                       -1    1 to 50 employees       -1            Unknown   
500                       -1                   -1       -1                 -1   
..                       ...                  ...      ...                ...   
209             Woodbine, MD    1 to 50 employees     2011  Company - Private   
405         Indianapolis, IN  51 to 200 employees     2016  Company - Private   
113           Burlingame, CA    1 to 50 employees     2016  Company - Private   
355             Woodbine, MD    1 to 50 employees     2011  Company - Private   
118  South San Francisco, CA  51 to 200 employees     2019  Company - Private   

                                    Industry                     Sector  \
457                                       -1                         -1   
425                               Consulting          Business Services   
351                                       -1                         -1   
504                                       -1                         -1   
500                                       -1                         -1   
..                                       ...                        ...   
209                              IT Services     Information Technology   
405  Enterprise Software & Network Solutions     Information Technology   
113             Computer Hardware & Software     Information Technology   
355                              IT Services     Information Technology   
118                Biotech & Pha

In [28]:
df[df['Rating']==-1.0].shape

(50, 20)

In [29]:
df.Rating = np.where(df.Rating==-1.0,0,df.Rating)

In [30]:
df.sort_values(by='Rating', ascending=True)

index                  Job Title Salary Estimate  \
457    457             data scientist        69K-$116   
425    425             data scientist       124K-$198   
351    351             data scientist       122K-$146   
504    504             data scientist        95K-$119   
500    500             data scientist        95K-$119   
..     ...                        ...             ...   
209    209             data scientist        79K-$106   
405    405  machine learning engineer       110K-$163   
113    113              data engineer        99K-$132   
355    355             data scientist       122K-$146   
118    118             data scientist        99K-$132   

                                       Job Description  Rating  \
457  Work in a fast growing startup with unlimited ...     0.0   
425  Job Description\nSelecting features, building ...     0.0   
351  About Our AI/ML Team\n\nOur mission is to buil...     0.0   
504  Job Description\nWorking at Sophinea\n\nSophin...     0.0   
500  Job Overview: The Data Scientist is a key memb...     0.0   
..                                                 ...     ...   
209  US Citizenship Required and (TS/SCI with FSP o...     5.0   
405  We are looking for an experienced engineer wit...     5.0   
113  At Phantom AI, experience the fast paced envir...     5.0   
355  US Citizenship Required and (TS/SCI with FSP o...     5.0   
118  ArsenalBio’s mission is to develop efficacious...     5.0   

                 Company Name                 Location  \
457             Stride Search        San Francisco, CA   
425              Microagility             New York, NY   
351          Point72 Ventures            Palo Alto, CA   
504                  Sophinea            Chantilly, VA   
500            Hatch Data Inc        San Francisco, CA   
..                        ...                      ...   
209  Phoenix Operations Group   Annapolis Junction, MD   
405                  LifeOmic              Raleigh, NC   
113                Phantom AI           Burlingame, CA   
355  Phoenix Operations Group   Annapolis Junction, MD   
118       Arsenal Biosciences  South San Francisco, CA   

                Headquarters                 Size  Founded  Type of ownership  \
457     Westlake Village, CA    1 to 50 employees       -1  Company - Private   
425            Princeton, NJ    1 to 50 employees       -1  Company - Private   
351                       -1                   -1       -1                 -1   
504                       -1    1 to 50 employees       -1            Unknown   
500                       -1                   -1       -1                 -1   
..                       ...                  ...      ...                ...   
209             Woodbine, MD    1 to 50 employees     2011  Company - Private   
405         Indianapolis, IN  51 to 200 employees     2016  Company - Private   
113           Burlingame, CA    1 to 50 employees     2016  Company - Private   
355             Woodbine, MD    1 to 50 employees     2011  Company - Private   
118  South San Francisco, CA  51 to 200 employees     2019  Company - Private   

                                    Industry                     Sector  \
457                                       -1                         -1   
425                               Consulting          Business Services   
351                                       -1                         -1   
504                                       -1                         -1   
500                                       -1                         -1   
..                                       ...                        ...   
209                              IT Services     Information Technology   
405  Enterprise Software & Network Solutions     Information Technology   
113             Computer Hardware & Software     Information Technology   
355                              IT Services     Information Technology   
118                Biotech & Pha

5- Location column - extracting abbreviation of the state

In [31]:
#Extracting job_state from Location Column
#Replacing Full Names with appropriate state code
#Remote and Utah deleting the rows
df['Location'].apply(lambda x: x.split(",")[-1]).value_counts()

 CA              165
 VA               89
 MA               62
 NY               52
 MD               40
 IL               30
 DC               26
 TX               17
 WA               16
 OH               14
 PA               12
 MO               12
United States     11
 NJ               10
 CO               10
 GA                9
 NC                9
 TN                8
 FL                8
Remote             6
 OK                6
 WI                6
 IN                5
 MI                5
 AL                4
 CT                4
 MN                4
 AZ                4
 NE                3
 UT                3
 IA                3
 RI                2
New Jersey         2
 SC                2
 OR                2
Utah               2
 NH                2
 MS                1
 LA                1
 KS                1
Texas              1
 DE                1
California         1
 WV                1
Name: Location, dtype: int64

In [32]:
df['job_state']=df['Location'].apply(lambda x: x.split(",")[-1].strip())
df['job_state'].replace(['United States','Texas','California','New Jersey','Remote','Utah'],["US","TX","CA","NJ",np.nan,np.nan],inplace=True)
df.dropna(axis=0,inplace=True)
df.reset_index(inplace=True)

In [33]:
df['job_state'].value_counts()

CA    166
VA     89
MA     62
NY     52
MD     40
IL     30
DC     26
TX     18
WA     16
OH     14
NJ     12
PA     12
MO     12
US     11
CO     10
GA      9
NC      9
FL      8
TN      8
WI      6
OK      6
IN      5
MI      5
MN      4
AL      4
CT      4
AZ      4
IA      3
UT      3
NE      3
SC      2
RI      2
OR      2
NH      2
MS      1
LA      1
KS      1
DE      1
WV      1
Name: job_state, dtype: int64

In [34]:
df.head()

level_0  index              Job Title Salary Estimate  \
0        0      0  senior data scientist       137K-$171   
1        1      1         data scientist       137K-$171   
2        2      2         data scientist       137K-$171   
3        3      3         data scientist       137K-$171   
4        4      4         data scientist       137K-$171   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

         Company Name       Location            Headquarters  \
0         Healthfirst   New York, NY            New York, NY   
1             ManTech  Chantilly, VA             Herndon, VA   
2      Analysis Group     Boston, MA              Boston, MA   
3             INFICON     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions   New York, NY            New York, NY   

                      Size  ...                               Industry  \
0   1001 to 5000 employees  ...                     Insurance Carriers   
1  5001 to 10000 employees  ...                 Research & Development   
2   1001 to 5000 employees  ...                             Consulting   
3    501 to 1000 employees  ...  Electrical & Electronic Manufacturing   
4      51 to 200 employees  ...                Advertising & Marketing   

              Sector                     Revenue  \
0          Insurance    Unknown / Non-Applicable   
1  Business Services      $1 to $2 billion (USD)   
2  Business Services  $100 to $500 million (USD)   
3      Manufacturing  $100 to $500 million (USD)   
4  Business Services    Unknown / Non-Applicable   

                                         Competitors salary_lower  \
0            EmblemHealth, UnitedHealth Group, Aetna          137   
1                                                 -1          137   
2                                                 -1          137   
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...          137   
4               Commerce Signals, Cardlytics, Yodlee          137   

  salary_upper salary_midpoint          old title  integrated title job_state  
0          171           154.0  Sr Data Scientist                 1        NY  
1          171           154.0     Data Scientist                 0        VA  
2          171           154.0     Data Scientist                 0        MA  
3          171           154.0     Data Scientist                 0        MA  
4          171           154.0     Data Scientist                 0        NY  

[5 rows x 22 columns]

6-Drop unwanted column

In [35]:
df.drop(['Founded','Competitors','level_0'], axis=1,inplace=True)

In [36]:
df

index              Job Title Salary Estimate  \
0        0  senior data scientist       137K-$171   
1        1         data scientist       137K-$171   
2        2         data scientist       137K-$171   
3        3         data scientist       137K-$171   
4        4         data scientist       137K-$171   
..     ...                    ...             ...   
659    667         data scientist       105K-$167   
660    668         data scientist       105K-$167   
661    669         data scientist       105K-$167   
662    670         data scientist       105K-$167   
663    671         data scientist       105K-$167   

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
659  Summary\n\nWe’re looking for a data scientist ...     3.6   
660  Job Description\nBecome a thought leader withi...     0.0   
661  Join a thriving company that is changing the w...     0.0   
662  100 Remote Opportunity As an AINLP Data Scient...     5.0   
663  Description\n\nThe Data Scientist will be part...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
659              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
660                  JKGT  San Francisco, CA                      -1   
661            AccessHope      Irwindale, CA                      -1   
662  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
663         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
659   1001 to 5000 employees        Company - Private   
660                       -1                       -1   
661                       -1                       -1   
662        1 to 50 employees        Company - Private   
663   1001 to 5000 employees         Company - Public   

                                  Industry             Sector  \
0                       Insurance Carriers          Insurance   
1                   Research & Development  Business Services   
2                               Consulting  Business Services   
3    Electrical & Electronic Manufacturing      Manufacturing   
4                  Advertising & Marketing  Business Services   
..                                     ...                ...   
659                Advertising & Marketing  Business Services   
660                                     -1                 -1   
661                                     -1                 -1   
662                Advertising & Marketing  Business Services   
663                              Wholesale  Business Services   

                        Revenue salary_lower salary_upper  salary_midpoint  \
0      Unknown / Non-Applicable          137          171            154.0   
1        $1 to $2 billion (USD)          137          171            154.0   
2    $100 

# **8. Conclusion**

Now we have the raw data cleaned, and the data is ready for the exploratory data analysis.

In [37]:
df.to_csv('Cleaned_DS_Jobs.csv', index=False)